platform/darwin/src/MGLOfflineStorage.mm

```diff
@@ -649,8 +658,14 @@ - (void)preloadData:(NSData *)data forURL:(NSURL *)url modificationDate:(nullabl
     if (expires) {
         response.expires = mbgl::Timestamp() + std::chrono::duration_cast<mbgl::Seconds>(MGLDurationFromTimeInterval(expires.timeIntervalSince1970));
     }
-    
-    _mbglDatabaseFileSource->put(resource, response);
+
+    if (completion) {
+        _mbglDatabaseFileSource->forward(resource, response, [completion] {
+            dispatch_async(dispatch_get_main_queue(), completion);
+        });
+    } else {
+        _mbglDatabaseFileSource->put(resource, response);
```

Comment:
Is `put()` still asynchronous? What’s the difference between calling `forward()` with a no-op lambda and calling `put()`?

Reply:
> What’s the difference between calling forward() with a no-op lambda and calling put()

`put` is still asynchronous. `forward` with callback was introduced to keep compatibility for an old `put` method.

Reply:
Since `put` doesn’t have any advantage over `forward`, should we mark the preexisting public method as deprecated, to make it easier to remove in the future?

Reply:
@1ec5 sure, we can plan deprecation on native side.